sinks/sink-common/src/connector/batching.rs

```diff
@@ -0,0 +1,173 @@
+use std::time::Instant;
+
+use apibara_core::node::v1alpha2::Cursor;
+use error_stack::Result;
+use serde_json::Value;
+use tracing::{info, warn};
+
+use crate::{error::SinkError, sink::Context, CursorAction};
+
+#[derive(Debug)]
+pub struct Buffer {
+    pub start_at: Instant,
+    pub start_cursor: Cursor,
+    pub end_cursor: Cursor,
+    pub data: Vec<Value>,
+}
+
+impl Default for Buffer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Buffer {
+    pub fn new() -> Self {
+        Self {
+            start_at: Instant::now(),
+            start_cursor: Cursor::default(),
+            end_cursor: Cursor::default(),
+            data: Vec::new(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn to_vec(&self) -> Vec<Value> {
+        self.data.to_vec()
+    }
+
+    pub fn extend(&mut self, data: Vec<Value>) {
+        self.data.extend(data);
+    }
+
+    pub fn clear(&mut self) {
+        self.start_at = Instant::now();
+        self.start_cursor = Cursor::default();
+        self.end_cursor = Cursor::default();
+        self.data.clear();
+    }
+}
+
+pub struct Batcher {
+    pub batch_size: u64,
+    pub batch_secs: u64,
+    pub buffer: Buffer,
+}
+
+impl Batcher {
+    pub fn by_secs(batch_secs: u64) -> Self {
+        Self {
+            batch_size: 0,
+            batch_secs,
+            buffer: Buffer::new(),
+        }
+    }
+
+    pub fn by_size(batch_size: u64) -> Self {
+        Self {
+            batch_size,
+            batch_secs: 0,
+            buffer: Buffer::new(),
+        }
+    }
+
+    pub fn is_batching_by_secs(&self) -> bool {
+        self.batch_secs != 0
+    }
+
+    pub fn is_batching_by_size(&self) -> bool {
+        self.batch_size != 0
+    }
+
+    pub fn is_batching(&self) -> bool {
+        self.is_batching_by_secs() || self.is_batching_by_size()
+    }
+
+    /// Check if the batch is already added to the buffer
+    pub fn is_already_added(&self, batch: &[Value]) -> bool {
```

Comment:
This operation can be expensive because `contains` on a list does a linear search. Let's keep it this way for now, but later we should fix it by tracking `context.cursor` from `handle_batch` and `handle_invalidate`.

Reply:
Added a TODO comment to not forget it